src/Member/MemberUtil.php

```diff
@@ -12,19 +12,26 @@
 use Contao\FilesModel;
 use Contao\Folder;
 use Contao\MemberModel;
+use Contao\StringUtil;
 use Contao\System;
 use Contao\Validator;
+use HeimrichHannot\UtilsBundle\Model\ModelUtil;
 
 class MemberUtil
 {
     /**
      * @var ContaoFrameworkInterface
      */
     protected $framework;
+    /**
+     * @var ModelUtil
+     */
+    protected $modelUtil;
 
-    public function __construct(ContaoFrameworkInterface $framework)
+    public function __construct(ContaoFrameworkInterface $framework, ModelUtil $modelUtil)
```

Comment:
This might produce circular reference issues. Please remove that one and use System::getContainer()->get(ModelUtil::class) in this special case (utils-bundle) instead.

Reply:
Frage: wieso? System sollten wir nicht mehr verwenden.

Reply:
Weil es circular references erzeugt, wenn du zufällig im ModelUtil auch das MemberUtil verwendest. Wird hier nicht passieren, ist theoretisch aber denkbar. Hatte ich im Drafts-Bundle bereits, da hat ein Service den anderen required und es hat geknallt.

Reply:
Klar, auf sowas muss man bei der Entwicklung achten. Aber dann wieder so eine krumme Stütze zu nutzen, da bin ich dagegen. Lieber einer Circular Exception risikieren, die einem sagt, dass man da Mist gebaut hat, also den Mist verstecken, bis das Kartenhaus einbricht und man es eh beheben muss. YMHO ;)

Reply:
OK, von mir aus. Aber das wird sich glaube ich nicht verhindern lassen. Wir nutzen die Utils aktuell auch schon querbeet hin und her (was auch vollkommen ok und kein "Mist" ist). Damit sparen wir uns doppelten Code. Wenn du mal *alle* System::getContainers gegen Autowiring austauschst, wird es garantiert irgendwo knallen.

Wie würdest du denn die *benötigte" zirkuläre Referenz beheben ohne den Container?

Reply:
Der Code muss eben so gebaut werden, dass es keine zirkulare Referenz gibt. Da gibt es kein Wenn und Aber. Wenn diese auftritt, muss der Code umgebaut werden, eventuell auch aufgespalten. Hier helfen natürlich klar Strukturen, bspw. sollten grundlegene Utils am besten keine andere Util-Abhängigkeit haben (bspw. String, Model), während es beim MemberUtil durchaus ok ist, auf das ModelUtil zurückzugreifen. Mit ein wenig Hirnschmalz und dem kritischen Blick der Kollegen sollte es schon klappen :)

Reply:
OK, das können wir als Leitlinie gern so machen. Grundlegende Dinge sollten für sich genommen funktionieren. Darauf aufbauende speziellere Strukturen wie Member können dann andere Utils nutzen.

Evtl. können wir es aber auch so auflösen, wenn mal eine zirkuläre Referenz auftritt?

https://stackoverflow.com/questions/49554736/symfony4-service-container-circular-dependency

Was meint ihr? Das braucht man im aktuellen Fall nicht (@AlexejKossmann Hier bitte wie besprochen wieder auf ModelUtil Autowiring im Konstruktor zurück), aber grundsätzlich wäre das eine Option so wie es aussieht. Leider hat der Beitrag nicht viel Aufmerksamkeit, weil das Problem nicht oft auftritt oder es noch andere populärere Threads gibt gibt.

Reply:
> OK, das können wir als Leitlinie gern so machen. Grundlegende Dinge sollten für sich genommen funktionieren. Darauf aufbauende speziellere Strukturen wie Member können dann andere Utils nutzen.

Klingt aus meiner Sicht gut.

Meinst du die Setter-Injection? Das wäre natürlich auch noch ein Option. Ich denke aber, durch ein gute Struktur lässt sich das in den meisten Fällen anders regeln.

Reply:
Ich glaube zirkuläre Referenzen löst man typischerweise auf, indem man sie in eine eigene Klasse verallgemeinert, d.h.

A braucht B.method1
B braucht A.method2

wird dann zu

A braucht C.method1
B braucht C.method2

bzw.

A braucht C.method1
B braucht D.method2

Das wird in der Praxis aber evtl. zu unnötig komplexem Code führen.

Schauen wir es am besten immer im Einzelfall an ;-)

Reply:
Richtig, hier würde ich im Einzelfall entscheiden. Aber ein wenig mehr aufspalten und abstrahieren wäre in diesem Bundle hier sogar manchmal ganz gut. Bspw. sind die 4 Methoden in diesem PR eigentlich 2, welche sich nur in der Tabelle unterscheiden. Würde man es ganz korrekt machen, müsste man eine generische Methode extrahieren :)

Reply:
Stimmt. Evtl. ein Trait?

Reply:
Ja, das könnte man hier so machen. "PersonTrait" oder so :D

Reply:
Klingt gut :-) @AlexejKossmann Kannst du das bitte bauen unter src/Traits/PersonTrait.php ? Dann die Methode verallgemeinern und in UserUtil und MemberUtil die Methoden parametrisiert aufrufen?